packages/flame/lib/src/experimental/layout_component.dart

```diff
@@ -0,0 +1,146 @@
+import 'package:flame/components.dart';
+import 'package:flame/experimental.dart';
+import 'package:flame/src/game/notifying_vector2.dart';
+import 'package:flutter/widgets.dart';
+
+enum Direction { horizontal, vertical }
+
+/// Super class for [RowComponent] and [ColumnComponent]
+/// A relayout is performed when
+///  - a change in this component's children takes place
+///  - the size of this component or a parent changes
+///  - the [gap] parameter is changed
+abstract class LayoutComponent extends PositionComponent {
+  LayoutComponent(
+    this.direction,
+    this.mainAxisAlignment,
+    this._gap,
+    Vector2? size,
+  )   : isManuallySized = size != null,
+        super(
+          size: size,
+        );
+  final Direction direction;
+  final MainAxisAlignment mainAxisAlignment;
+  bool isManuallySized;
+
+  /// gap between components
+  double _gap;
+
+  set gap(double gap) {
+    _gap = gap;
+    layoutChildren();
+  }
+
+  double get gap => _gap;
+
+  NotifyingVector2? get _bounds =>
+      isManuallySized ? size : findParent<PositionComponent>()?.size;
+
+  @override
+  Future<void> onLoad() async {
+    final size = _bounds;
+    size?.addListener(layoutChildren);
+  }
+
+  @override
+  void onChildrenChanged(Component child, ChildrenChangeType changeType) {
+    layoutChildren();
+  }
+
+  void layoutChildren() {
+    final list = children.whereType<PositionComponent>().toList();
+    if (list.isEmpty) {
+      return;
+    }
+    final currentPosition = Vector2.zero();
+    final componentsDimension = list.fold<double>(
+      0,
+      (previousValue, element) => direction == Direction.horizontal
+          ? previousValue + element.width
+          : previousValue + element.height,
+    );
+    final vectorIndex = direction == Direction.horizontal ? 0 : 1;
+    final gapDimension = gap * (list.length - 1);
+
+    /// Either we have a defined size or we just use the summed width/height of
+    /// the current children.
+    final componentBounds = _bounds;
+    final dimensionAvailable =
```

Comment:
Most of the variables in this method are confusingly named. When naming variables, I find it easier to first try to explain what the variable represents, and then convert the first few words of that explanation into a name.

For example, consider `componentsDimension`. An explanation for this variable would sound like "the total width/height of all children components". So the variable name could be something like `totalSizeOfComponents`. Similarly, `gapDimension` -> `totalGapsSize`, `dimensionAvailable` -> `availableSpace`, `freeSpacePerComponent` -> `extraGapBetweenComponents`, etc. Another good synonym for "dimension" is "extent".